File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/cozy/CozyMarkerElementsBuilder.java
package io.flutter.plugins.googlemaps.cozy;

import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.RectF;
import android.graphics.Rect;
import android.graphics.Typeface;

import com.caverock.androidsvg.SVG;

import java.util.regex.Pattern;
import java.util.regex.Matcher;

import io.flutter.plugins.googlemaps.MarkerCache;

class CozyMarkerElementsBuilder {
    private final Typeface font;
    private final MarkerCache markerCache;
    private final float strokeSize;

    public CozyMarkerElementsBuilder(Typeface font, MarkerCache markerCache, float strokeSize) {
        this.font = font;
        this.markerCache = markerCache;
        this.strokeSize = strokeSize;
    }

    public static float getDpFromPx(float px) {
        return px * Resources.getSystem().getDisplayMetrics().density;
    }

    private float getTextYOffset(float markerHeight, Rect rect) {
        return (markerHeight / 2f) + (rect.height() / 2f) - rect.bottom;
    }

    private float getTextXOffset(float markerWidth, Rect rect) {
        return (markerWidth / 2f) - (rect.width() / 2f) - rect.left;
    }

    private Bitmap getIconBitmap(String svgIcon, int width, int height, int rgbColor) {
        if (svgIcon == null)
            return null;

        String key = String.format("%d%d%d%d", svgIcon.hashCode(), width, height, rgbColor);

        if (key == null)
            return null;

        final Bitmap bitmap = markerCache.getBitmapFromMemCache(key);
        if (bitmap != null) {
            return bitmap;
        }

        try {
            //Needed this because svg doesn't scale if there is fixed width and height
            String widthPattern = "(width=\")(.+?)(\")";
            String heightPattern = "(height=\")(.+?)(\")";

            Matcher widthMatcher = Pattern.compile(widthPattern).matcher(svgIcon);
            Matcher heightMatcher = Pattern.compile(heightPattern).matcher(svgIcon);

            String finalSvgIcon = svgIcon.replaceAll(widthPattern, "");
            finalSvgIcon = finalSvgIcon.replaceAll(heightPattern, "");

            finalSvgIcon = recolorSvg(finalSvgIcon, rgbColor);

            SVG svg = SVG.getFromString(finalSvgIcon);
            if (widthMatcher.find() && heightMatcher.find()) {
                svg.setDocumentViewBox(0, 0, Integer.parseInt(widthMatcher.group(2)), Integer.parseInt(heightMatcher.group(2)));
            }

            Bitmap iconBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
            Canvas canvas = new Canvas(iconBitmap);
            svg.renderToCanvas(canvas, new RectF(0, 0, width, height));

            markerCache.addBitmapToMemoryCache(svgIcon, iconBitmap);
            return iconBitmap;
        } catch (Exception e) {
        }

        return null;
    }

    private String recolorSvg(String originalSvg, int rgbColor) {
        String hexColor = String.format("#%06X", (0xFFFFFF & rgbColor));
        return originalSvg.replaceAll("(fill=\")(.+?)(\")", "fill=\"" + hexColor + "\"");
    }

    public CozyMarkerElements cozyElementsFromData(CozyMarkerData markerData) {
        String text = markerData.label;
        String counterText = markerData.counter;

        boolean hasPointer = markerData.hasPointer;
        boolean hasCounter = counterText != null && !counterText.isEmpty();

        /* setting colors */
        final int defaultMarkerColor = Color.WHITE;
        final int defaultTextColor = Color.BLACK;
        final int defaultStrokeColor = Color.rgb(217, 219, 208);
        final int defaultIconCircleColor = Color.rgb(248, 249, 245);
        final int defaultIconColor = Color.BLACK;

        final int defaultCounterBubbleColor = Color.rgb(235, 237, 230);

        final int selectedMarkerColor = Color.rgb(57, 87, 189);
        final int selectedTextColor = Color.WHITE;
        final int selectedIconCircleColor = Color.WHITE;
        final int selectedCounterBubbleColor = Color.rgb(248, 249, 245);

        final int visualizedMarkerColor = Color.rgb(217, 219, 208);
        final int visualizedTextColor = Color.BLACK;
        final int visualizedStrokeColor = Color.rgb(197, 201, 186);
        final int visualizedIconCircleColor = Color.WHITE;
        final int visualizedCounterBubbleColor = Color.rgb(248, 249, 245);

        final int specialIconCircleColor = Color.rgb(240, 243, 255);
        final int specialIconColor = Color.rgb(57, 87, 189);

        int markerColor = defaultMarkerColor;
        int textColor = defaultTextColor;
        int counterTextColor = defaultTextColor;
        int strokeColor = defaultStrokeColor;
        int iconCircleColor = defaultIconCircleColor;
        int iconColor = defaultIconColor;
        int counterBubbleColor = defaultCounterBubbleColor;

        if (markerData.isVisualized) {
            markerColor = visualizedMarkerColor;
            textColor = visualizedTextColor;
            strokeColor = visualizedStrokeColor;
            iconCircleColor = visualizedIconCircleColor;
            counterBubbleColor = visualizedCounterBubbleColor;
        }
        if (markerData.isSelected) {
            markerColor = selectedMarkerColor;
            textColor = selectedTextColor;
            strokeColor = defaultStrokeColor;
            iconCircleColor = selectedIconCircleColor;
            counterBubbleColor = selectedCounterBubbleColor;
        }
        if (markerData.variant.equals("special") &&
                (!markerData.isVisualized || markerData.isSelected)) {
            iconCircleColor = specialIconCircleColor;
            iconColor = specialIconColor;
        }

        /* setting constants */
        // setting global constants
        final float paddingVertical = getDpFromPx(12);
        final float paddingHorizontal = getDpFromPx(11.5f);
        final float minMarkerWidth = getDpFromPx(40);
        final float strokeSize = getDpFromPx(1.5f);

        // setting constants for counter
        final float counterBubblePadding = getDpFromPx(6);

        // setting constants for pointer
        final float pointerWidth = getDpFromPx(7);
        final float pointerHeight = getDpFromPx(6);

        // setting constants for icon
        final float iconSize = getDpFromPx(16);
        final float iconCircleSize = getDpFromPx(24);
        final float iconLeftPadding = getDpFromPx(5);
        final float iconRightPadding = getDpFromPx(3);

        /* setting variables */
        // gets the text size based on the font
        Rect textBounds = new Rect();
        Rect counterTextBounds = new Rect();

        float textSize = getDpFromPx(12f);
        Paint textPaint = new Paint();
        textPaint.setTypeface(font);
        textPaint.setTextSize(textSize);
        textPaint.setAntiAlias(true);
        textPaint.setTextAlign(Paint.Align.LEFT);

        textPaint.getTextBounds(text, 0, text.length(), textBounds);
        if (hasCounter) {
            textPaint.getTextBounds(counterText, 0, counterText.length(), counterTextBounds);
        }

        // getting icon bitmap
        Bitmap iconBitmap = getIconBitmap(markerData.icon, (int) Math.round(iconSize),
                (int) Math.round(iconSize), iconColor);

        // set the marker height as the string height with space for padding and stroke
        float markerHeight = textBounds.height() + (2f * paddingVertical) + 2f * strokeSize;

        // bubble shape properties which used in marker width
        float bubbleShapeHeight = markerHeight - strokeSize;
        float counterBubbleShapeHeight = bubbleShapeHeight - (2f * counterBubblePadding);

        // counterSummaryWidth to be used on markerWidth
        float counterSummaryWidth = hasCounter ?
                Math.max(counterTextBounds.width() + (2f * counterBubblePadding),
                        counterBubbleShapeHeight) : 0;

        // additionalIconWidth to be used on markerWidth
        float iconAdditionalWidth = iconBitmap != null ? iconCircleSize + iconRightPadding : 0;

        // pointerSize to be used on bitmap creation
        float pointerSize = (hasPointer ? pointerHeight : 0);

        // setting marker width with a minimum width in case the string size is below the minimum
        float markerWidth = textBounds.width() + counterSummaryWidth + (2f * paddingHorizontal)
                + (2f * strokeSize) + iconAdditionalWidth;
        if (markerWidth < minMarkerWidth) {
            markerWidth = minMarkerWidth;
        }

        // marker bubble coordinates
        float bubbleShapeX = strokeSize / 2f;
        float bubbleShapeY = strokeSize / 2f;
        float bubbleShapeWidth = markerWidth - strokeSize;

        // counter bubble coordinates
        float counterBubbleShapeX = markerWidth - strokeSize
                - counterSummaryWidth - counterBubblePadding;
        float counterBubbleShapeY = counterBubblePadding + strokeSize / 2f;

        // counter text coordinates and size
        float counterTextWidth = counterTextBounds.width();
        float counterTextHeight = textSize;
        float counterTextDx = counterBubbleShapeX +
                (counterSummaryWidth / 2f) - (counterTextWidth / 2f);
        float counterTextDy = getTextYOffset(markerHeight, counterTextBounds);

        // other important variables
        float middleOfMarkerY = (bubbleShapeHeight / 2f) + strokeSize / 2f;

        // text coordinates
        float textDx = getTextXOffset(markerWidth, textBounds) + iconAdditionalWidth / 2f
                - counterSummaryWidth / 2f;
        float textDy = getTextYOffset(markerHeight, textBounds);
        float textHeight = textSize;
        float textWidth = textBounds.width();

        // Icon coordinates
        float iconX = iconLeftPadding + strokeSize + (iconCircleSize - iconSize) / 2f;
        float iconY = (bubbleShapeHeight / 2f) - (iconSize / 2f) + strokeSize / 2f;
        float iconWidth = iconSize;
        float iconHeight = iconSize;

        // Icon circle coordinates
        float circleX = iconLeftPadding + strokeSize + iconCircleSize / 2f - iconCircleSize / 2f;
        float circleY = middleOfMarkerY - iconCircleSize / 2f;
        float circleWidth = iconCircleSize;
        float circleHeight = iconCircleSize;

        // Pointer coordinates
        float pointerX = markerWidth / 2f - pointerWidth;
        float pointerY = markerHeight - strokeSize;

        return new CozyMarkerElements(
                // Canvas
                new CozyMarkerElement(
                        new RectF(0, 0, markerWidth, markerHeight + pointerSize)
                ),
                // Bubbles
                new CozyMarkerElement(
                        new RectF(bubbleShapeX, bubbleShapeY, bubbleShapeX + bubbleShapeWidth, bubbleShapeY + bubbleShapeHeight),
                        markerColor,
                        strokeColor
                ),
                // Counter bubble
                new CozyMarkerElement(
                        hasCounter ? new RectF(counterBubbleShapeX, counterBubbleShapeY, counterBubbleShapeX + counterSummaryWidth, counterBubbleShapeY + counterBubbleShapeHeight) : null,
                        counterBubbleColor,
                        defaultStrokeColor
                ),
                // Labels
                new CozyMarkerElement[]{
                        new CozyMarkerElement(
                                new RectF(textDx, textDy, textDx + textWidth, textDy + textHeight),
                                textColor,
                                1.0f,
                                text
                        ),
                        new CozyMarkerElement(
                                new RectF(counterTextDx, counterTextDy, counterTextDx + counterTextWidth, counterTextDy + counterTextHeight),
                                counterTextColor,
                                1.0f,
                                counterText
                        ),
                },
                // Icon
                new CozyMarkerElement[]{
                        new CozyMarkerElement(
                                new RectF(iconX, iconY, iconX + iconWidth, iconY + iconHeight),
                                iconColor,
                                iconBitmap == null ? 0.0f : 1.0f,
                                iconBitmap == null ? null : iconBitmap
                        ),
                },
                // IconCircle
                new CozyMarkerElement(
                        new RectF(circleX, circleY, circleX + circleWidth, circleY + circleHeight),
                        iconCircleColor,
                        iconBitmap == null ? 0.0f : 1.0f
                ),
                // Pointer
                new CozyMarkerElement(
                        new RectF(pointerX, pointerY, pointerX + 2 * pointerWidth, pointerY + pointerSize),
                        markerColor,
                        strokeColor
                )
        );
    }
}

File: packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/CozyMarkerElementsBuilder.m
//
//  CozyMarkerElementsBuilder.m
//  Pods
//
//  Created by Pietro Domingues on 26/09/23.
//

#import <Foundation/Foundation.h>
#import <CoreText/CoreText.h>
#import "CozyMarkerElementsBuilder.h"
#import "CozyMarkerData.h"
#import "CozyMarkerElements.h"

@implementation CozyMarkerElementsBuilder


- (instancetype)initWithFontPath:(NSString *)fontPath strokeSize:(CGFloat)strokeSize {
    self = [super init];
    if (self) {
        _fontPath = fontPath;
        _strokeSize = strokeSize;
    }
    return self;
}

- (CozyMarkerElements *)cozyElementsFromData:(CozyMarkerData *)cozyMarkerData {
    NSString *text = cozyMarkerData.label;
    NSString *icon = cozyMarkerData.icon;
    NSString *counterText = cozyMarkerData.counter;
    BOOL hasPointer = cozyMarkerData.hasPointer;
    
    /* setting colors */
    UIColor *const defaultMarkerColor = UIColor.whiteColor;
    UIColor *const defaultTextColor = UIColor.blackColor;
    UIColor *const defaultStrokeColor = [UIColor colorWithRed:
                                         217.0f / 255.0f
                                                        green:(219.0f /
                                                               255.0f) blue:(
                                                                             208.0f /
                                                                             255.0f)                             alpha:1];
    
    UIColor *const defaultIconCircleColor = [UIColor colorWithRed:(
                                                                   248.0f / 255.0f)
                                                            green:(249.f /
                                                                   255.0f) blue:(
                                                                                 245.0f /
                                                                                 255.0f)                                 alpha:1];

    UIColor *const counterBubbleColor = [UIColor colorWithRed:(
                                                               235.0f / 255.0f)
                                                        green:(237.f /
                                                               255.0f) blue:(
                                                                             230.0f /
                                                                             255.0f)                             alpha:1];
    
    UIColor *const defaultIconColor = UIColor.blackColor;
    
    UIColor *const selectedMarkerColor = [UIColor colorWithRed:(57.0f / 255.0f) green:(87.0f /
                                                                                       255.0f) blue:(
                                                                                                     189.0f / 255.0f) alpha:1];
    UIColor *const selectedTextColor = UIColor.whiteColor;
    UIColor *const selectedIconCircleColor = UIColor.whiteColor;
    
    UIColor *const visualizedMarkerColor = [UIColor colorWithRed:(217.0f / 255.0f) green:(219.0f /
                                                                                          255.0f) blue:(
                                                                                                        208.0f / 255.0f) alpha:1];
    UIColor *const visualizedTextColor = UIColor.blackColor;
    UIColor *const visualizedStrokeColor = [UIColor colorWithRed:(197.0f / 255.0f) green:(201.0f / 255.0f) blue:(186.0f / 255.0f) alpha:1];
    UIColor *const visualizedIconCircleColor = UIColor.whiteColor;
    
    UIColor *const specialIconCircleColor = [UIColor colorWithRed:(240.0f / 255.0f) green:(243.0f /
                                                                                           255.0f) blue:(
                                                                                                         255.0f / 255.0f) alpha:1];
    UIColor *const specialIconColor = [UIColor colorWithRed:(57.0f / 255.0f) green:(87.0f /
                                                                                    255.0f) blue:(
                                                                                                  189.0f / 255.0f) alpha:1];
    
    UIColor *markerColor = defaultMarkerColor;
    UIColor *textColor = defaultTextColor;
    UIColor *iconCircleColor = defaultIconCircleColor;
    UIColor *iconColor = defaultIconColor;
    UIColor *strokeColor = defaultStrokeColor;
    
    if (cozyMarkerData.isVisualized) {
        markerColor = visualizedMarkerColor;
        textColor = visualizedTextColor;
        strokeColor = visualizedStrokeColor;
        iconCircleColor = visualizedIconCircleColor;
    }
    if (cozyMarkerData.isSelected) {
        markerColor = selectedMarkerColor;
        textColor = selectedTextColor;
        strokeColor = defaultStrokeColor;
        iconCircleColor = selectedIconCircleColor;
    }
    if ([cozyMarkerData.variant isEqualToString:@"special"] &&
        (!cozyMarkerData.isVisualized || cozyMarkerData.isSelected)) {
        iconCircleColor = specialIconCircleColor;
        iconColor = specialIconColor;
    }
    
    /* setting constants */
    // setting padding and stroke size
    const CGFloat paddingVertical = 10.5f;
    const CGFloat paddingHorizontal = 11;
    const CGFloat minMarkerWidth = 40;
    
    // setting constants for pointer
    const CGFloat pointerWidth = 6;
    const CGFloat pointerHeight = 5;
    
    // setting constants for icon
    const CGFloat iconSize = 16;
    const CGFloat iconCircleSize = 24;
    const CGFloat iconLeftPadding = 5;
    const CGFloat iconRightPadding = 3;
    
    // setting constants for coutner
    const CGFloat counterBubblePadding = 6;
    
    /* setting variables */
    // getting font and setting its size to 3 the size of the marker size
    const CGFloat fontSize = 12;
    UIFont *textFont = [UIFont fontWithName:self.fontPath size:fontSize];
    CGSize stringSize = [text sizeWithAttributes:@{NSFontAttributeName: textFont}];
    CGSize counterTextSize = [counterText sizeWithAttributes:@{NSFontAttributeName: textFont}];
    
    // additionalIconWidth to be used on markerWidth
    CGFloat iconAdditionalWidth = icon != NULL ? iconCircleSize + iconRightPadding : 0;
    
    // set the marker height as the string height with space for padding and stroke
    CGFloat markerHeight = stringSize.height + (2 * paddingVertical) + 2 * self.strokeSize;
    
    // bubbles shape properties
    CGFloat bubbleShapeHeight = markerHeight - self.strokeSize * 2;
    CGFloat counterBubbleShapeHeight = bubbleShapeHeight - (counterBubblePadding * 2);
    
    // counter summary width
    CGFloat counterSummaryWidth = counterText ? MAX(counterTextSize.width + counterBubblePadding * 2,
                                                    counterBubbleShapeHeight) : 0;
    
    // pointerSize to be used on bitmap creation
    CGFloat pointerSize = hasPointer ? pointerHeight : 0;
    
    // setting marker width with a minimum width in case the string size is below the minimum
    CGFloat markerWidth = stringSize.width + (2 * paddingHorizontal) + (2 * self.strokeSize) +
    iconAdditionalWidth + counterSummaryWidth;
    if (markerWidth < minMarkerWidth) {
        markerWidth = minMarkerWidth;
    }
    
    // bubble coordinates
    CGFloat bubbleShapeX = self.strokeSize;
    CGFloat bubbleShapeY = self.strokeSize;
    CGFloat bubbleShapeWidth = markerWidth - self.strokeSize * 2;
    
    // counter bubble coordinates
    CGFloat counterBubbleShapeX =
    markerWidth - counterSummaryWidth - counterBubblePadding - self.strokeSize;
    CGFloat counterBubbleShapeY = counterBubblePadding + self.strokeSize;
    
    // generic position variables
    CGFloat middleOfMarkerY = (bubbleShapeHeight / 2) + self.strokeSize;
    
    // counter text coordinates
    CGFloat counterTextX =
    counterBubbleShapeX + (counterSummaryWidth / 2) - (counterTextSize.width / 2);
    CGFloat counterTextY = middleOfMarkerY - (counterTextSize.height / 2);
    
    // text coordinates
    CGFloat textY = middleOfMarkerY - (stringSize.height / 2);
    CGFloat textX = (markerWidth / 2) - (stringSize.width / 2) + iconAdditionalWidth / 2 - counterSummaryWidth / 2;
    CGFloat textWidth = stringSize.width;
    CGFloat textHeight = stringSize.height;
    
    // icon coordinates
    CGFloat iconX = self.strokeSize + iconLeftPadding + (iconCircleSize - iconSize) / 2;
    CGFloat iconY = middleOfMarkerY - iconSize / 2;
    CGFloat iconWidth = iconSize;
    CGFloat iconHeight = iconSize;
    
    // icon circle coordinates
    CGFloat iconCircleX = self.strokeSize + iconLeftPadding;
    CGFloat iconCircleY = middleOfMarkerY - iconCircleSize / 2;
    CGFloat iconCircleWidth = iconCircleSize;
    CGFloat iconCircleHeight = iconCircleSize;
    
    // pointer coordinates
    CGFloat pointerX = markerWidth / 2 - pointerWidth;
    CGFloat pointerY = markerHeight - self.strokeSize;
    
    return [[CozyMarkerElements alloc] initWithCanvas:[[CozyMarkerElement alloc]
                                                       initWithBounds:CGRectMake(0, 0, markerWidth, markerHeight + pointerSize)
                                                       fillColor:markerColor
                                                       strokeColor:strokeColor
                                                       alpha:1
                                                       data:nil]
                                               bubble:[[CozyMarkerElement alloc]
                                                       initWithBounds:CGRectMake(bubbleShapeX,
                                                                                 bubbleShapeY,
                                                                                 bubbleShapeWidth,
                                                                                 bubbleShapeHeight)
                                                       fillColor:markerColor
                                                       strokeColor:strokeColor
                                                       alpha:1
                                                       data:nil]
                                               labels:@[[[CozyMarkerElement alloc]
                                                         initWithBounds:CGRectMake(textX,
                                                        textY,
                                                        textWidth,
                                                        textHeight)
                                                         fillColor:textColor
                                                         strokeColor:nil
                                                         alpha:1
                                                         data:text],
                                                        [[CozyMarkerElement alloc]
                                                         initWithBounds:CGRectMake(
                                                        counterTextX,
                                                        counterTextY,
                                                        counterTextSize.width,
                                                        counterTextSize.height)
                                                         fillColor:defaultTextColor
                                                         strokeColor:nil
                                                         alpha:1
                                                         data:counterText]]
                                              counter:[[CozyMarkerElement alloc]
                                                       initWithBounds:CGRectMake(counterBubbleShapeX,
                                                                                 counterBubbleShapeY,
                                                                                 counterSummaryWidth,
                                                                                 counterBubbleShapeHeight)
                                                       fillColor:counterBubbleColor
                                                       strokeColor:strokeColor
                                                       alpha:1
                                                       data:counterText]
                                                 icon:[[CozyMarkerElement alloc]
                                                       initWithBounds:CGRectMake(iconX, iconY,
                                                                                 iconWidth,
                                                                                 iconHeight)
                                                       fillColor:iconColor
                                                       strokeColor:nil
                                                       alpha:icon == NULL ? 0 : 1
                                                       data:icon]
                                           iconCircle:[[CozyMarkerElement alloc]
                                                       initWithBounds:CGRectMake(iconCircleX,
                                                                                 iconCircleY,
                                                                                 iconCircleWidth,
                                                                                 iconCircleHeight)
                                                       fillColor:iconCircleColor
                                                       strokeColor:nil
                                                       alpha:icon == NULL ? 0 : 1
                                                       data:nil]
                                              pointer:[[CozyMarkerElement alloc]
                                                       initWithBounds:CGRectMake(pointerX, pointerY,
                                                                                 2 * pointerWidth,
                                                                                 pointerSize)
                                                       fillColor:markerColor
                                                       strokeColor:markerColor
                                                       alpha:1
                                                       data:nil]];
    
}

@end
